### PR TITLE
W-15633176: add initial AVRO Schema parsing

### PR DESCRIPTION
**Pull request by arielmirra**
- **chore: refactor buildConfig cycle test method**
- **W-15633189: add IsAVROSchema Eternal annotation**
- **chore: refactor JsonSchemaDocumentParser**
- **W-15633184: add AVRO fields to AnyShape**
- **W-15633184: add AvroSchemaDocument (model and instances)**
- **W-15633184: add Avro schema parsing**

### CODE REVIEW

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/domain/AsyncSchemaEmitter.scala [hunk 1]
```diff
@@ -53,5 +54,16 @@ case class AsyncSchemaEmitter(
     )
   }
 
+  private def emitAsAvro(b: EntryBuilder, schemaVersion: SchemaVersion, avroType: String): Unit = {
+    b.entry(
+      key,
+      b => {
+        val newCtx = new Async20SpecEmitterContext(spec.eh, config = spec.renderConfig, schemaVersion = schemaVersion)
+        OasTypePartEmitter(shape, ordering, references = references)(OasLikeShapeEmitterContextAdapter(newCtx))
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/domain/AsyncSchemaEmitter.scala:63]: This should call a specific AVRO Schema emitter (starting point for issue W-15633198). Add a TODO to not forget it.

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroEnumParser.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import amf.core.internal.parser.{YMapOps, YScalarYRead}
+import amf.shapes.client.scala.model.domain.AnyShape
+import amf.shapes.internal.domain.metamodel.AnyShapeModel
+import org.yaml.model._
+
+class AvroEnumParser(map: YMap)(implicit ctx: AvroSchemaContext) extends AvroScalarShapeParser("string", Some(map)) {
+
+  override def parse(): AnyShape = {
+    val shape = super.parse()
+    map.key("name", (AnyShapeModel.Name in shape).allowingAnnotations)
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroEnumParser.scala]: "name", "namespace", "aliases" and "doc" are parsed in multiple shapes (Enum, Record, Fixed) the same way. They could be extracted to a trait or similar.

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroFixedShapeParser.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import amf.shapes.client.scala.model.domain.{AnyShape, ScalarShape}
+import amf.shapes.internal.domain.metamodel.AnyShapeModel
+import org.yaml.model.YMap
+
+case class AvroFixedShapeParser(map: YMap)(implicit ctx: AvroSchemaContext) extends AvroShapeBaseParser(map) {
+
+  override def parse(): AnyShape = {
+    val datatype = Xsd.base + "fixed" // todo: is this correct? necessary? same question as scalar shape parser
+    val shape    = ScalarShape(Annotations(map)).withDataType(datatype, Annotations(map.entries.head))
+    map.key("name", AnyShapeModel.Name in shape)
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroFixedShapeParser.scala]: Same comment to extract common fields

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroRecordParser.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import amf.core.internal.parser.domain.Annotations
+import amf.shapes.client.scala.model.domain.{AnyShape, NodeShape}
+import amf.shapes.internal.domain.metamodel.AnyShapeModel
+import org.yaml.model.{YMap, YMapEntry, YNode}
+
+class AvroRecordParser(map: YMap)(implicit ctx: AvroSchemaContext) extends AvroShapeBaseParser(map) {
+  private val shape = NodeShape(map)
+
+  override def parse(): NodeShape = {
+    map.key("name", (AnyShapeModel.Name in shape).allowingAnnotations)
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroRecordParser.scala]: Same comment to extract common fields

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroScalarShapeParser.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  private val (annotations, typeAnnotations): (Annotations, Annotations) =
+    maybeMap.map(annotationsFromMap).getOrElse(defaultAnnotations)
+
+  private def annotationsFromMap(map: YMap) = (Annotations(map), Annotations(map.entries.head))
+  private def nilShape                      = NilShape(annotations).withName(`type`)
+  private def scalarShape = ScalarShape(annotations).withName(`type`).withDataType(dataType, typeAnnotations)
+  private def dataType: String = {
+    // todo: bytes is ok as xsd#bytes? or shall we use format?
+    // todo: maybe use TypeDef? investigate map to see if it has the info inside, validate that is a valid AVRO primitive type
+    XsdTypeDefMapping.xsdFromString(`type`)._1.getOrElse(Xsd.base + `type`)
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroScalarShapeParser.scala]: As we talk last week, Scalar parsing should be done using the TypeDef already provided by SYAML, similar to what happens with the JSON/OAS Scalar parser.

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroShapeParser.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  private def parseType(name: String): AnyShape = {
+    name match {
+      case "map"                 => parseMap()
+      case "array"               => parseArray()
+      case "record"              => parseRecord()
+      case "enum"                => parseEnum()
+      case "fixed"               => parseFixed()
+      case _ if name.isPrimitive => parsePrimitiveType(name)
+      case _                     => AnyShape() // ignore todo: validate UnsuppportedTypeValidation()
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroShapeParser.scala]: We should discuss what to do with the parsing errors. My first thought was to leave all the validation to the specific AVRO Schema library (when we add it, in the validation epic) but if we miss information the emitted AVRO Schema will miss the broken parts... Maybe save the raw schema in a non eternal annotation for validation is a good solution...

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/common/reference/ApiReferenceHandler.scala [hunk 7]
```diff
@@ -85,6 +85,7 @@ class ApiReferenceHandler(spec: String) extends ReferenceHandler {
       case Raml10.id | Raml08.id             => ramlLinks(part)
       case Oas20.id | Oas30.id | AwsOas30.id => oasLinks(part)
       case JsonSchema.id                     => oasLinks(part)
+      case AvroSchema.id                     => oasLinks(part) // TODO: what to put here?
```
Comment (looseale) [inline amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/common/reference/ApiReferenceHandler.scala]: I think, this is for links from the Spec (AVRO in this case) to another. AVRO Schemas cannot have refs, so it should never enter here.

File: amf-shapes/shared/src/main/scala/amf/shapes/internal/annotations/Annotations.scala [hunk 8]
```diff
@@ -15,6 +15,15 @@ object ParsedJSONSchema extends AnnotationGraphLoader {
     Some(ParsedJSONSchema(value))
 }
 
+case class IsAVROSchema(avroType: String) extends EternalSerializedAnnotation {
```
Comment (looseale) [inline amf-shapes/shared/src/main/scala/amf/shapes/internal/annotations/Annotations.scala]: I would change the name of this to "AVROSchemaType" or something more descriptive

**Review by looseale — CHANGES_REQUESTED** [review]

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroShapeParser.scala (hunk 6 above)
Comment (looseale) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroShapeParser.scala]: No change required here, is only to not forget it

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroEnumParser.scala (hunk 2 above)
Comment (arielmirra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroEnumParser.scala:15]: done

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/domain/AsyncSchemaEmitter.scala (hunk 1 above)
Comment (arielmirra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/async/emitters/domain/AsyncSchemaEmitter.scala:63]: done

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroScalarShapeParser.scala (hunk 5 above)
Comment (arielmirra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroScalarShapeParser.scala]: Done

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/common/reference/ApiReferenceHandler.scala (hunk 7 above)
Comment (arielmirra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/common/reference/ApiReferenceHandler.scala]: great, removed 👌🏽

File: amf-shapes/shared/src/main/scala/amf/shapes/internal/annotations/Annotations.scala (hunk 8 above)
Comment (arielmirra) [reply amf-shapes/shared/src/main/scala/amf/shapes/internal/annotations/Annotations.scala]: Done 👌🏽

File: amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroShapeParser.scala (hunk 6 above)
Comment (arielmirra) [reply amf-api-contract/shared/src/main/scala/amf/apicontract/internal/spec/avro/parser/domain/AvroShapeParser.scala]: Great!

**Review by looseale — APPROVED** [review]

**Pull request merged**